File: apps/api/aqapi/level_tables/entity/allowance_by_level_entity.py
from sqlalchemy import Column, Integer, ForeignKey, DateTime, CheckConstraint, UniqueConstraint

from sqlalchemy.orm import relationship, Mapped, mapped_column
from sqlalchemy.sql import func
from aqapi.core.entity.base_entity import BaseEntity
from aqapi.core.entity.base_translation_entity import BaseTranslationEntity
from aqapi.core.config.db_config import DB_CONF


class AllowanceByLevelEntity(BaseEntity):
    """レベル別お小遣いテーブルエンティティ"""

    __tablename__ = "allowance_by_level"
    __table_args__ = (
        # レベルは0以上
        CheckConstraint("level >= 0", name="chk_allowance_by_level_level_positive"),
        # 金額は0以上
        CheckConstraint("amount >= 0", name="chk_allowance_by_level_amount_positive"),
        # 一意制約
        UniqueConstraint("level", "allowance_table_id", name="uq_allowance_by_level"),
    )

    allowance_table_id: Mapped[int] = mapped_column(Integer, ForeignKey("allowance_tables.id", ondelete="CASCADE"), nullable=False, comment="お小遣いテーブルID")
    level: Mapped[int] = mapped_column(Integer, nullable=False, comment="レベル")
    amount: Mapped[int] = mapped_column(Integer, nullable=False, default=0, comment="お小遣い金額")

    allowance_tables = relationship("AllowanceTablesEntity", foreign_keys=[allowance_table_id])

File: packages/backend/src/features/level-tables/entity/allowanceByLevelEntity.ts
import {
  Entity,
  Column,
  ManyToOne,
  JoinColumn,
  Check,
  Unique,
} from "typeorm";
import { BaseTransactionEntity } from "@backend/core/entity/baseTransactionEntity";
import { AllowanceTableEntity } from "src/features/allowance-table/entity/allowanceTableEntity";

/**
 * レベル別お小遣いテーブルエンティティ
 */
@Entity("allowance_by_level")
@Check("chk_allowance_by_level_level_positive", "level >= 0")
@Check("chk_allowance_by_level_amount_positive", "amount >= 0")
@Unique("uq_allowance_by_level", ["level", "allowance_table_id"])
export class AllowanceByLevelEntity extends BaseTransactionEntity {
  @Column({ type: "int", nullable: false, comment: "お小遣いテーブルID" })
  allowance_table_id!: number;
  @Column({ type: "int", nullable: false, comment: "レベル" })
  level!: number;
  @Column({ type: "int", nullable: false, default: 0, comment: "お小遣い金額" })
  amount!: number;
  
  @ManyToOne(() => AllowanceTableEntity, { nullable: false, onDelete: "CASCADE" })
  @JoinColumn({ name: "allowance_table_id", referencedColumnName: "id", foreignKeyConstraintName: "fk_allowance_by_level_allowance_table_id" })
  allowance_table!: AllowanceTableEntity;
}
